src/processor.rs

```diff
@@ -208,4 +214,70 @@ impl Processor {
 
 		Ok(())
 	}
+
+	/// Cancel can be called after init_escrow. If called after exchange, it's already too late
+	/// since exchange is atomic and tokens have been transferred, so there will be no effect.
+	/// Since tokens haven't actually been transferred from initializer main token account to
+	/// initializer temp token account, we don't need to actually transfer any tokens.
+	/// What we need to do is:
+	/// 1) Close escrow info
+	/// 2) Close PDA account
+	fn process_cancel(accounts: &[AccountInfo], amount_expected_to_return: u64, program_id: &Pubkey) -> ProgramResult {
```

Comment:
exchange and cancel have quite a bit of functional overlap and the code should reflect that to avoid code repetition (and bugs - see below).